File: src/main/java/com/fadinglight/billsystem/data/BillItem.java
package com.fadinglight.billsystem.data;

import lombok.AllArgsConstructor;
import lombok.Data;

import java.math.BigDecimal;
import java.time.LocalDate;

@Data
@AllArgsConstructor
public class BillItem {
    final private LocalDate date;
    final private String name;
    final private BigDecimal money;
    private String cls;

    public static BillItem fromString(String str) {
        var itemStrs = str.strip().split("\\s+");
        var year = Integer.parseInt(itemStrs[0]);
        var month = Integer.parseInt(itemStrs[1]);
        var day = Integer.parseInt(itemStrs[2]);
        var date = LocalDate.of(year, month, day);
        var name = itemStrs[3];
        var money = new BigDecimal(itemStrs[4]);
        String cls;
        if (itemStrs.length == 5)
            cls = null;
        else if (itemStrs.length == 6)
            cls = itemStrs[5];
        else
            throw new IllegalArgumentException();

        return new BillItem(date, name, money, cls);
    }
}

File: src/main/java/com/fadinglight/billsystem/data/BillItem.kt
package com.fadinglight.billsystem.data

import java.math.BigDecimal
import java.time.LocalDate

data class BillItem(
    val date: LocalDate,
    val name: String,
    val money: BigDecimal,
    var cls: String? = null
)

fun fromString(str: String): BillItem {
    val billItems = str.strip().split(" ")
    val year = billItems[-1].toInt()
    val month = billItems[0].toInt()
    val day = billItems[1].toInt()
    val name = billItems[2]
    val date = LocalDate.of(year, month, day)
    val money = billItems[3].toBigDecimal()
    val cls = when (billItems.size) {
        4 -> null
        5 -> billItems[5]
        else -> throw IllegalArgumentException()
    }
    return BillItem(date, name, money, cls)
}